modules/swagger-parser-v3/src/main/java/io/swagger/parser/v3/util/OpenAPIDeserializer.java

```diff
@@ -1,23 +1,2081 @@
+            if (array != null) {
+                openAPI.setSecurity(getSecurityRequirementsList(array, "security", result));
+            }
+            openAPI.setExtensions(new LinkedHashMap<>());
+
+            Set<String> keys = getKeys(rootNode);
+            for(String key : keys) {
+                if(key.startsWith("x-")) {
+                    openAPI.getExtensions().put(key, Json.mapper().convertValue(rootNode.get(key), Object.class));
+                }
+                else if(!ROOT_KEYS.contains(key)) {
+                    result.extra(location, key, rootNode.get(key));
+                }
+            }
+
+        } else {
+            result.invalidType("", "", "object", node);
+            result.invalid();
+            return null;
+        }
+
+
+        return openAPI;
+    }
+
+    public Components getComponents(ObjectNode obj, String location, ParseResult result) {
+        if (obj == null) {
+            return null;
+        }
+        Components components = new Components();
+
+        ObjectNode node = getObject("schemas",obj,false,location,result);
+        if(node != null) {
+            components.setSchemas(getSchemas(node,location,result));
+        }
+
+        node = getObject("responses",obj,false,location,result);
+        if(node != null) {
+            components.setResponses(getResponses(node,location,result));
+        }
+
+        node = getObject("parameters",obj,false,location,result);
+        if(node != null) {
+            components.setParameters(getParameters(node,location,result));
+        }
+        node = getObject("examples",obj,false,location,result);
+        if(node != null) {
+            components.setExamples(getExamples(node,location,result));
+        }
+
+        node = getObject("requestBodies",obj,false,location,result);
+        if(node != null) {
+            components.setRequestBodies(getRequestBodies(node,location,result));
+        }
+
+        node = getObject("headers",obj,false,location,result);
+        if(node != null) {
+            components.setHeaders(getHeaders(node,location,result));
+        }
+
+        node = getObject("securitySchemes",obj,false,location,result);
+        if(node != null) {
+            components.setSecuritySchemes(getSecuritySchemes(node,location,result));
+        }
+
+        node = getObject("links",obj,false,location,result);
+        if(node != null) {
+            components.setLinks(getLinks(node,location,result));
+        }
+
+        node = getObject("callbacks",obj,false,location,result);
+        if(node != null) {
+            components.setCallbacks(getCallbacks(node,location,result));
+        }
+        components.setExtensions(new LinkedHashMap<>());
+
+        Set<String> keys = getKeys(obj);
+        for(String key : keys) {
+            if(key.startsWith("x-")) {
+                components.getExtensions().put(key, Json.mapper().convertValue(obj.get(key), Object.class));
```

Comment:
I suggest moving the extraction of extensions into a single method.  I'm not sure if we want to return `Object.class`--if so, we will have just a single location to change.  This occurs in multiple places.